synapse/api/auth.py

```diff
@@ -173,6 +173,20 @@ async def get_user_by_req(
                     parent_span.set_tag("device_id", requester.device_id)
                 if requester.app_service is not None:
                     parent_span.set_tag("appservice_id", requester.app_service.id)
+
+                # Tag any headers that we need to extract from the request. This
+                # is useful to specify any headers that a reverse-proxy in front
+                # of Synapse may be sending to correlate and match up something
+                # in that layer to a Synapse trace. ex. when Cloudflare times
+                # out it gives a `cf-ray` header which we can also tag here to
+                # find the trace.
+                for header_key in self.hs.config.tracing.request_headers_to_tag:
+                    headers = request.requestHeaders.getRawHeaders(header_key)
+                    if headers is not None:
+                        parent_span.set_tag(
+                            SynapseTags.REQUEST_HEADER_PREFIX + header_key, headers[0]
+                        )
```

Comment:
I've tested that this works locally but I'm not sure if the `cf-ray` that we want to use on `matrix.org` is actually a thing.

Based on the following documentation and that we [see it in our haproxy logs](https://github.com/matrix-org/matrix-ansible-private/issues/2689) already, I assume this will be viable,

> Add the [`CF-Ray` header to your origin web server logs](https://support.cloudflare.com/hc/en-us/articles/203118044#h_f7a7396f-ec41-4c52-abf5-a110cadaca7c) to match requests proxied to Cloudflare to requests in your server logs.